crates/bevy_input_focus/src/directional_navigation.rs

```diff
@@ -0,0 +1,379 @@
+
+use bevy_app::prelude::*;
+use bevy_ecs::{
+    entity::{EntityHashMap, EntityHashSet},
+    prelude::*,
+    system::SystemParam,
+};
+use bevy_math::CompassOctant;
+use thiserror::Error;
+
+use crate::InputFocus;
+
+/// A plugin that sets up the directional navigation systems and resources.
+#[derive(Default)]
+pub struct DirectionalNavigationPlugin;
+
+impl Plugin for DirectionalNavigationPlugin {
+    fn build(&self, app: &mut App) {
+        app.init_resource::<DirectionalNavigationMap>();
+    }
+}
+
+/// The up-to-eight neighbors of a focusable entity, one for each [`CompassOctant`].
+#[derive(Default, Debug, Clone, PartialEq)]
+pub struct NavNeighbors {
+    /// The array of neighbors, one for each [`CompassOctant`].
+    /// The mapping between array elemnents and directions is determined by [`CompassOctant::to_index`].
+    ///
+    /// If no neighbor exists in a given direction, the value will be [`None`].
+    /// In most cases, using [`NavNeighbors::set`] and [`NavNeighbors::get`]
+    /// will be more ergonomic than directly accessing this array.
+    pub neighbors: [Option<Entity>; 8],
+}
+
+impl NavNeighbors {
+    /// An empty set of neighbors.
+    pub const EMPTY: NavNeighbors = NavNeighbors {
+        neighbors: [None; 8],
+    };
+
+    /// Get the neighbor for a given [`CompassOctant`].
+    pub const fn get(&self, octant: CompassOctant) -> Option<Entity> {
+        self.neighbors[octant.to_index()]
+    }
+
+    /// Set the neighbor for a given [`CompassOctant`].
+    pub const fn set(&mut self, octant: CompassOctant, entity: Entity) {
+        self.neighbors[octant.to_index()] = Some(entity);
+    }
+}
+
+/// A resource that stores the traversable graph of focusable entities.
+///
+/// Each entity can have up to 8 neighbors, one for each [`CompassOctant`].
+///
+/// To ensure that your graph is intuitive to navigate and generally works correctly, it should be:
+///
+/// - **Connected**: Every focusable entity should be reachable from every other focusable entity.
+/// - **Symmetric**: If entity A is a neighbor of entity B, then entity B should be a neighbor of entity A, ideally in the reverse direction.
+/// - **Physical**: The direction of navigation should match the layout of the entities when possible,
+///   although looping around the edges of the screen is also acceptable.
+/// - **Not self-connected**: An entity should not be a neighbor of itself; use [`None`] instead.
+///
+/// For now, this graph must be built manually, and the developer is responsible for ensuring that it meets the above criteria.
+#[derive(Resource, Debug, Default, Clone, PartialEq)]
+pub struct DirectionalNavigationMap {
+    /// A directed graph of focusable entities.
+    ///
+    /// Pass in the current focus as a key, and get back a collection of up to 8 neighbors,
+    /// each keyed by a [`CompassOctant`].
+    pub neighbors: EntityHashMap<NavNeighbors>,
```

Comment:
Just a thought, could it make sense for `NavNeighbors` to be a component? Possibly an immutable one?

Reply:
Yeah, I chewed on this a fair bit. IMO until we have relations we shouldn't try to encode these graphs via components.